### PR TITLE
[Wav2Vec2] Improve Tokenizer & Model for batched inference

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR improves batched inference for Wav2Vec2 models by:

- adding an `attention_mask`
- adding zero-mean unit-variance normalization to the tokenizer
- correctly setting returning `attention_mask` and doing normalization depending on which architecture is used

## Background

Some of Fairseq's Wav2Vec2 models apply Group Normalization over the time axis in the feature extractor. This means that the convolutional layers in the feature extractor can not 100% correctly treat padded input resulting in those models giving different results depending on whether the input is padded or not. See https://github.com/pytorch/fairseq/issues/3227 . Those models should never make use of `attention_mask` which is made sure by setting `return_attention_mask=False` in their corresponding tokenizer configs: https://huggingface.co/facebook/wav2vec2-base-960h/blob/main/tokenizer_config.json . Also some explicit warnigs have been added to both the tokenizer and model.

For the "newer" models however that have the improved layer norm architecture in the feature extraction: https://huggingface.co/facebook/wav2vec2-large-960h-lv60-self , normalization and correct padding via `attention_mask` gives some nice performance improvements and works correctly.

## Performance Evaluation

 I've evaluated both `wav2vec2-large-960h-lv60-self` and `wav2vec2-large-960h-lv60` on the test set of librispeech and got some nice improvements:

- `wav2vec2-large-960h-lv60-self`: 2.2 WER -> 1.8 WER
- `wav2vec2-large-960h-lv60`: 3.4 WER -> 2.2 WER 

So that the results now seem to match the paper's results very nicely.

Also, I checked that `wav2vec2-base-960h` should **not** use an `attention_mask` as the performance on librispeech test then drop heavily from ~4 WER to ~20 WER.

## TODO

Once this PR is merged, I can fully focus on adding the fine-tuning functionality and will also update the model cards with the new evaluation code & results.

### CODE REVIEW

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -67,6 +67,7 @@ def forward(self, hidden_states):
 class Wav2Vec2LayerNormConvLayer(nn.Module):
     def __init__(self, config, layer_id=0):
         super().__init__()
+        self.id = layer_id
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:70]:
```suggestion
```

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 2]
```diff
@@ -191,7 +191,6 @@ def __init__(self, config):
         self.dropout = nn.Dropout(config.feat_extract_dropout)
 
     def forward(self, hidden_states):
-        hidden_states = hidden_states.transpose(1, 2)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:194]: this operation was badly placed here and was moved further downward

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 3]
```diff
@@ -544,6 +573,19 @@ def _init_weights(self, module):
         if isinstance(module, (nn.Linear, nn.Conv1d)) and module.bias is not None:
             module.bias.data.zero_()
 
+    def _get_feat_extract_output_lengths(self, input_lengths: torch.LongTensor):
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:576]: This function is quite important for speech since there is often a mismatch in sequence input length and model output length due to the use of `stride` in the convolutional layers. However, the model needs to know the "correct" feature extraction output length when treating padded input.

File: tests/test_tokenization_wav2vec2.py [hunk 4]
```diff
@@ -299,3 +301,46 @@ def test_tokenizer_slow_store_full_signature(self):
...
+        # wav2vec2-lv60 -> return attention_mask
+        tokenizer = self.get_tokenizer(return_attention_mask=True)
+        processed = tokenizer(speech_inputs, padding="longest")
+
+        self.assertIn("attention_mask", processed)
+        self.assertListEqual(list(processed.attention_mask.shape), list(processed.input_values.shape))
+        self.assertListEqual(processed.attention_mask.sum(-1).tolist(), [800, 1000, 1200])
+
+    @slow
+    def test_pretrained_checkpoints_are_set_correctly(self):
```
Comment (patrickvonplaten) [inline tests/test_tokenization_wav2vec2.py:336]: This test makes sure that the models configs are correctly set.

**Comment by youssefavx** [comment]
Go Patrick!!!!! YES! someone who cares!

**Review by sgugger — APPROVED** [review]
Looks good to me, great docstrings!

**Comment by patrickvonplaten** [comment]
Merging since @LysandreJik is off today and this is blocking me a bit
